sigma-dev/sigma
# Run Nomad Jobs on datacenters, not nodes

### changelog 

- [x] a serviceOperation now has a single `model.Site` instead of a slice 
- [x] if a Site has children, manager chooses one at random (including the parent in the pool) to deploy on 
- [x] tasker.Create now takes in a single Site instead of a slice 
- [x] in `createNomadJob`, removed the node constraint. No longer defaulting the datacenter to dc-local, but instead going into site attributes and grabbing `datacenter` string. This means that all services must provide a site name-- 
- [x] -- because there is no longer any site `local`. 
- [x] fixed up the nomad integration tests -- since we no longer have `tasker.Status`, now they just test `Create`, `isRunning`, and `Remove.`
- [x] added a multi-datacenter vagrant setup, `multi-dc`. Sets up 2 datacenters, `dc-1` and `dc2`, each with 2 clients. All fall under the same server, `server-1`. This lets you have multiple sites in sigma, each pointing to different datacenters, so that you can run services on different sites. 

### how to test 

1. stand up multiple client and server VMs with `./vagrant/nomad/multi-node`. You should be able to do a `nomad node-status` from the server and see `x` client nodes, where `x` is NUM_CLIENTS. Ssh into the clients and watch the logs-- this is how you can see which client gets the allocation to run the plugin container. Unfortunately just doing a `nomad alloc-status` won't tell you which client the Server sent it to. 
2. add a site where `Attributes["datacenter"] = "dc-local"` (see ./examples/helloWorld/addSite.sh) 
3. stand up sigma, add a plugin 
4. add a service and go into your client vms to watch the magic happen 

(If you want to be crazy you can also use `multi-dc`and have sites with `dc-1` and `dc-2` respectively.) 

### todo 

add customization for `multi-dc`
